File: benchmarks/cross-language/cpp/counting.cpp
// C++ Counting Actor Benchmark (Savina-style)
#include <iostream>
#include <iomanip>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <chrono>
#include <atomic>
#include <cstdlib>

static int MESSAGES = 100000;  // Default for "low" preset

class Counter {
    std::mutex mtx;
    std::condition_variable cv;
    int pending = 0;
    bool done = false;
    int processed = 0;

public:
    void increment() {
        std::unique_lock<std::mutex> lock(mtx);
        pending++;
        cv.notify_one();
    }

    void finish() {
        std::unique_lock<std::mutex> lock(mtx);
        done = true;
        cv.notify_one();
    }

    void run() {
        while (true) {
            std::unique_lock<std::mutex> lock(mtx);
            cv.wait(lock, [this] { return pending > 0 || done; });

            if (done && pending == 0) break;

            while (pending > 0) {
                pending--;
                processed++;
            }
        }
    }

    int getProcessed() const { return processed; }
};

int main() {
    const char* env = getenv("BENCHMARK_MESSAGES");
    if (env) MESSAGES = atoi(env);

    std::cout << "=== C++ Counting Actor Benchmark ===" << std::endl;
    std::cout << "Messages: " << MESSAGES << "\n" << std::endl;

    Counter counter;
    std::thread worker([&counter] { counter.run(); });

    auto start = std::chrono::high_resolution_clock::now();

    for (int i = 0; i < MESSAGES; i++) {
        counter.increment();
    }
    counter.finish();

    worker.join();

    auto end = std::chrono::high_resolution_clock::now();
    double elapsed = std::chrono::duration<double>(end - start).count();

    if (counter.getProcessed() != MESSAGES) {
        std::cout << "VALIDATION FAILED: expected " << MESSAGES
                  << ", got " << counter.getProcessed() << std::endl;
    }

    double throughput = MESSAGES / elapsed / 1e6;
    double ns_per_msg = elapsed * 1e9 / MESSAGES;

    std::cout << "ns/msg:         " << std::fixed << std::setprecision(2) << ns_per_msg << std::endl;
    std::cout << "Throughput:     " << std::setprecision(2) << throughput << " M msg/sec" << std::endl;

    return 0;
}

File: benchmarks/cross-language/scala/counting.scala
// Scala Akka Counting Actor Benchmark (Savina-style)
import akka.actor.{Actor, ActorRef, ActorSystem, Props}
import scala.concurrent.duration._
import scala.concurrent.{Await, Promise}

case object Increment
case class GetCount(replyTo: ActorRef)
case class Count(value: Long)

class CounterActor extends Actor {
  var count = 0L

  def receive = {
    case Increment =>
      count += 1
    case GetCount(replyTo) =>
      replyTo ! Count(count)
  }
}

class CollectorActor(promise: Promise[Long]) extends Actor {
  def receive = {
    case Count(value) =>
      promise.success(value)
      context.system.terminate()
  }
}

object CountingBenchmark extends App {
  val messages = sys.env.get("BENCHMARK_MESSAGES").flatMap(s => scala.util.Try(s.toLong).toOption).getOrElse(100000L)
  println("=== Scala Akka Counting Actor Benchmark ===")
  println(s"Messages: $messages\n")

  val system = ActorSystem("Counting")
  val promise = Promise[Long]()

  val counter = system.actorOf(Props[CounterActor]())
  val collector = system.actorOf(Props(classOf[CollectorActor], promise))

  val startTime = System.nanoTime()

  // Send all messages
  for (_ <- 0L until messages) {
    counter ! Increment
  }

  // Get count
  counter ! GetCount(collector)

  val count = Await.result(promise.future, 120.seconds)
  val endTime = System.nanoTime()

  val elapsed = (endTime - startTime).toDouble / 1e9

  if (count != messages) {
    println(s"VALIDATION FAILED: expected $messages, got $count")
  }

  val throughput = messages / elapsed / 1e6
  val nsPerMsg = elapsed * 1e9 / messages

  println(f"ns/msg:         $nsPerMsg%.2f")
  println(f"Throughput:     $throughput%.2f M msg/sec")

  Await.result(system.whenTerminated, 10.seconds)
}
